CPS-0003/README.md

```diff
@@ -0,0 +1,141 @@
+
+Both of these options would strain wallets and dapps, as they would need additional specific logic to handle smart tokens. The token validation scripts might not be generic enough and might be costly to do in every interaction. As the blockchain evolves, some interactions couldn't be predicted by the token and evolving the scripts would be problematic.
+
+A workaround might be, if some smart-token policies delegated the validation to some validations scripts. Unfortunately, without adding the ability to execute validation on referenced scripts, even this would require either: minting/forging a token in the transaction or spending an utxo with the desired validation logic. This approach doesn't seem scalable-enough, either.
+
+### Identifying smart tokens
+
+Drawing parallels with addresses, on cardano addresses have a 1 byte header identifying if they are a script address, pubkey address, reward address, etc. Unfortunately this is not the case with tokens currently. Policy IDs unfortunately do not have headers and it's assumed that it's always the hash of a native or plutus script.
+
+The identification is necessary, so that the nodes knows when it would require an additional transfer script to be present in the transaction. There might be 3 directions how "smart" or "validated" tokens could be identified in transactions:
+
+**1. segregate tokens**
+In this case, the validated and smart tokens could be separated in the transaction structures. In the CDDL:
+
+```cddl
+value = coin / [coin,multiasset<uint>] / [coin, multiasset<uint>, smartasset<uint>]
+mint = multiasset<int64> / [multiasset<uint>, smartasset<uint>]
+```
+
+Alternatively the value could be "objectified" similar to babbage tx outputs:
+
+```cddl
+value = {
+  0: coin,
+  ? 1: multiasset<uint>,
+  ? 2: smartasset<uint>
+}
+```
+
+This explicit segregation would have the benefit that it could be easier backwards compatible. In this format also the policy ids could remain as `hash28` script hashes.
+
+**2. add policy headers**
+This would be a breaking change to the structures and policies. Basically it would be redefining the multiasset structure to:
+
+```cddl
+multiasset<a> = { * policy_id_with_header => { * asset_name => a } }
+; token identifier format:
+; [ 8 bit header | policy_id ];
+; 0 - open transfer token
+; 1 - smart token
+policy_id_with_header = bytes
+policy_id = scripthash
+```
+
+**3. policy id length**
+
+In this situation, similar to how value can be just a coin or an array, the `policy_id` could be redefiend to be:
+
+```cddl
+policy_id = scripthash / bytes ; as above [ 8 bit header | scripthash ]
+```
+
+Both 2. and 3. are relatively error-prone, as many projects and tools assume 28byte policy IDs. Especially when encoding them as `subject` - for example for the offchain registry, there would be no way to tell, if the policyID is with the header or not.
+
+**TBD** add more considerations
```

Comment:
I believe this section can be removed as the author has not actively maintained it and it should be its own CIP that seeks to address one or more of the open questions listed below.

Reply:
@Crypto2099 then let's not merge this until we decide what to do with the omitted material, to ensure we don't lose track of it... and let's talk about that at the meeting coming up this hour.

Reply:
Addressed with https://github.com/cardano-foundation/CIPs/pull/947/commits/5d2369cfe0385d7483478137418426b805110ebb as documented in https://github.com/cardano-foundation/CIPs/issues/969.